File: backend/src/crypto/providers/kdf.py
"""Grade A: HKDF-SHA3-256 key derivation."""

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from src.crypto.interfaces import KeyDerivationFunction


class HKDFSha3256KeyDerivationFunction(KeyDerivationFunction):
    def derive(self, shared_secret: bytes, *, info: bytes, length: int) -> bytes:
        hkdf = HKDF(algorithm=hashes.SHA3_256(), length=length, salt=None, info=info)
        return hkdf.derive(shared_secret)

File: frontend/src/crypto/providers/kdf.ts
/** Grade A: HKDF-SHA3-256 key derivation, via `@noble/hashes`. */

import { hkdf } from "@noble/hashes/hkdf";
import { sha3_256 } from "@noble/hashes/sha3";
import type { KeyDerivationFunction } from "../interfaces";

export const hkdfSha3256KeyDerivationFunction: KeyDerivationFunction = {
  async derive(sharedSecret, info, length) {
    return hkdf(sha3_256, sharedSecret, undefined, info, length);
  },
};
